### ISSUE TITLE
[BUG] SIGFPE when check_fraud() fuction is called

### ISSUE THREAD

**Issue by AdrianFretwell** [OpenSIPS/opensips#2712]
**OpenSIPS version you are running**

```
version: opensips 3.2.3 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
main.c compiled on  with gcc 10

```

**Describe the bug**

During an intial INVITE and after a Dialog has been created the routing script makes a call to check_fraud().  The precise code is given below:
    xlog("L_NOTICE", "FD: $var(t_fd_user) - $rU - $var(t_fd_p)");

    check_fraud($var(t_fd_user), $rU, $var(t_fd_p));

If a matching rule is found, then a SIGFPE is generated and OpenSIPS restarts, full log entry from start up to shutdown given further down the page.

**To Reproduce**
Steps to reproduce the behaviour:

1. Make sure at least one rule is in the fraud_detection table that will be matched by the INVITE 
2. Start OpenSIPS
2. Send INVITE
3. Call fails -> Inspect log output

**Expected behaviour**

check_fraud() will return 1 (matching rule found - call OK)

**Relevant System Logs**

Dec 17 11:59:22 outlaw opensips: INFO:core:fix_poll_method: using epoll as the IO watch method (auto detected)
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: NOTICE:core:main: version: opensips 3.2.3 (x86_64/linux)
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: NOTICE:core:main: using 1024 MB of shared memory, allocator: F_MALLOC
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: NOTICE:core:main: using 128 MB of private process memory, allocator: F_MALLOC
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:init_reactor_size: reactor size 262144 (using up to 10.00Mb of memory per process)
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_CORE_THRESHOLD(0)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_CORE_SHM_THRESHOLD(1)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_CORE_PKG_THRESHOLD(2)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_CORE_PROC_AUTO_SCALE(3)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:tm:mod_init: TM - initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:sl:mod_init: Initializing StateLess engine
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: NOTICE:signaling:mod_init: initializing module ...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:rr:mod_init: rr - initializing
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:maxfwd:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:sipmsgops:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:textops:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:db_postgres:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:usrloc:ul_init_locks: locks array size 512
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_UL_AOR_INSERT(4)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_UL_AOR_DELETE(5)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_UL_CONTACT_INSERT(6)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_UL_CONTACT_UPDATE(7)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_UL_CONTACT_DELETE(8)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_UL_CONTACT_REFRESH(9)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_UL_LATENCY_UPDATE(10)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:registrar:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:dialog:mod_init: Dialog module - initializing
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_DLG_STATE_CHANGED(11)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:acc:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_ACC_EVENT(12)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_ACC_CDR(13)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_ACC_MISSED_EVENT(14)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:auth:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:auth_db:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:alias_db:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:uac:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_RTPPROXY_STATUS(15)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:rtpproxy:rtpp_init_extra_stats: 0 42 [ ttl npkts_ina npkts_ino nrelayed ndropped]
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:rtpproxy:rtpp_init_extra_stats: 1 55 [ rtpa_nsent rtpa_nrcvd rtpa_ndups rtpa_nlost rtpa_perrs]
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_RTPPROXY_DTMF(16)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:dialplan:mod_init: initializing module...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:drouting:dr_init: dynamic routing - initializing
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:drouting:init_prefix_tree: counted 10 possible chars under a node
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_DROUTING_STATUS(17)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:permissions:parse_config_file: file not found: /etc/opensips/permissions.allow
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:permissions:mod_init: default allow file (/etc/opensips/permissions.allow) not found => empty rule set
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:permissions:parse_config_file: file not found: /etc/opensips/permissions.deny
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:permissions:mod_init: default deny file (/etc/opensips/permissions.deny) not found => empty rule set
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:avpops:avpops_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:exec:mod_init: exec - initializing
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:fraud_detection:mod_init: Initializing module
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_FRD_WARNING(18)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:evi_publish_event: Registered event <E_FRD_CRITICAL(19)>
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:tracer:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:options:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:topology_hiding:mod_init: initializing...
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266739]: INFO:core:mod_init: initializing UDP-plain protocol
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266754]: INFO:mi_datagram:datagram_process: a new child 0/266754
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266754]: INFO:rtpproxy:connect_rtpproxies: created to 5
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266756]: INFO:mi_datagram:datagram_process: a new child 1/266756
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266758]: INFO:rtpproxy:connect_rtpproxies: created to 94
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266756]: INFO:rtpproxy:connect_rtpproxies: created to 5
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266754]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266756]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266758]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266766]: INFO:rtpproxy:connect_rtpproxies: created to 94
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266766]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266771]: INFO:rtpproxy:connect_rtpproxies: created to 94
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266768]: INFO:rtpproxy:connect_rtpproxies: created to 94
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266768]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266771]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266774]: INFO:rtpproxy:connect_rtpproxies: created to 94
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266774]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266772]: INFO:rtpproxy:connect_rtpproxies: created to 94
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266764]: INFO:rtpproxy:connect_rtpproxies: created to 94
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266773]: INFO:rtpproxy:connect_rtpproxies: created to 94
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266770]: INFO:rtpproxy:connect_rtpproxies: created to 94
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266772]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266764]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266773]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266770]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266765]: INFO:rtpproxy:connect_rtpproxies: created to 94
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266767]: INFO:rtpproxy:connect_rtpproxies: created to 94
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266765]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266767]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266769]: INFO:rtpproxy:connect_rtpproxies: created to 94
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266769]: INFO:rtpproxy:rtpp_test: rtp proxy <udp:22.33.44.44:12221> found, support for it enabled
Dec 17 11:59:22 outlaw opensips: INFO:core:daemonize: pre-daemon process exiting with 0
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266764]: INFO:drouting:dr_reload_data_head: loading drouting data!
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266764]: NOTICE:drouting:dr_load_routing_info: loaded 7 gateways in partition 'Default'
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266764]: NOTICE:drouting:dr_load_routing_info: loaded 0 carriers in partition 'Default'
Dec 17 11:59:22 outlaw /usr/sbin/opensips[266764]: NOTICE:drouting:dr_load_routing_info: loaded 4 rules in partition 'Default'
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266765]: traffic_route: called sip:OP0022@xxxxxxx.co.uk:5060 sip:07770759924@xxxxxxx.co.uk:5060 from 233.109.109.109:15064 method INVITE on 1.2.3.4
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266765]: FD: OP0022 - 07770759924 - 2
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:handle_sigs: child process 266765 exited by a signal 8
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:handle_sigs: core was not generated
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:handle_sigs: terminating due to SIGCHLD
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266763]: INFO:core:sig_usr: signal 15 received
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266759]: INFO:core:sig_usr: signal 15 received
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266761]: INFO:core:sig_usr: signal 15 received
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 14(266770) [SIP receiver udp:4.3.2.1:5060] terminated, still waiting for 16 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 8(266764) [SIP receiver udp:1.2.3.4:5060] terminated, still waiting for 15 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 1(266753) [RTPP notification receiver] terminated, still waiting for 14 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 4(266758) [MI FIFO] terminated, still waiting for 13 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 5(266759) [Stun loop] terminated, still waiting for 12 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 2(266754) [MI Datagram] terminated, still waiting for 11 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 3(266756) [MI Datagram] terminated, still waiting for 10 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 11(266767) [SIP receiver udp:1.2.3.4:5060] terminated, still waiting for 9 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 7(266763) [timer] terminated, still waiting for 8 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 10(266766) [SIP receiver udp:1.2.3.4:5060] terminated, still waiting for 7 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 13(266769) [SIP receiver udp:4.3.2.1:5060] terminated, still waiting for 6 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 16(266772) [SIP receiver udp:1.2.3.4:8060] terminated, still waiting for 5 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 17(266773) [SIP receiver udp:1.2.3.4:8060] terminated, still waiting for 4 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 12(266768) [SIP receiver udp:4.3.2.1:5060] terminated, still waiting for 3 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 15(266771) [SIP receiver udp:4.3.2.1:5060] terminated, still waiting for 2 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 18(266774) [Timer handler] terminated, still waiting for 1 more
Dec 17 12:02:42 outlaw /usr/sbin/opensips[266739]: INFO:core:shutdown_opensips: process 6(266761) [time_keeper] terminated, still waiting for 0 more

**OS/environment information**
 - Operating System: Debian 11
 - OpenSIPS installation: deb https://apt.opensips.org bullseye 3.2-releases
 - other relevant information: database module db_postgres

**Additional context**

**Comment by AdrianFretwell** [comment]
Using GDB (gdb /usr/local/sbin/opensips -c /core)  I have managed to get a little more information:

Core was generated by `/usr/local/sbin/opensips -P /run/opensips/opensips.pid -f /usr/local/etc/opensi'.
Program terminated with signal SIGFPE, Arithmetic exception.
#0  0x000055555567661c in check_freq_interval (_atp=<optimized out>, _trp=0x7fffb118b880) at time_rec.c:1266
1266				return (((_t1-_t0)/(7*24*3600))%_trp->interval==0)?

**Comment by liviuchircu** [comment]
Hello, @AdrianFretwell!  And thanks for the nice report!  This looks to be an issue on versions `3.2` and above, since the time recurrence engine was not only enhanced, but also curated of several bugs, such as the impossibility to use the `0` timestamp (UNIX start time), as it was considered to be the default field value (not set).  The solution was to change the default initialization of `0` to `-1`, which cannot be a timestamp anymore.

Unfortunately, it looks like `fraud_detection` is one of few modules which are still building these time recurrence structures using the old API, rather than just passing a string blob and letting the engine do the parsing -- I did not consider this case at all!

TL;DR: could you apply below patch on `3.2`+ by placing yourself in the OpenSIPS root source directory then copy-pasting the patch command into your console, recompile, retest and let me know if it works for you?  Thank you in advance!

```
git apply <(base64 -d <<EOF | gzip -dc
H4sIAAAAAAAAA7VXbXPaRhD+DL9inU6DBAgkXgSEkNh1PG2mTuwhbsedNqMR0iloRi9UOlpwnP/e
3dMrMmDamX7RoX2722f3Vg+26zigKF9cDmbXD+21x+KuE5lr27AZZxZ3wwDfbcMLTbtjweIEo7ob
2GwD/YGzGPXYZNTpLEbOcDLW7QloqqoPBnVFUU7ar95qtU7b8/wclN5g1NahRcsIUBBzk7sWuAEH
K2ImZwZ3fWZEzJKsMIg5GkTQFDI0jXgbqmIW2O061Hzmx4xLPNq0QW1D7D6w0EGTaCNPUY2r8saN
DY+ZDsxAQ1m9VeM+bmS4gUuOzEotmaW8sdeRSedHWwl3MJbQhL6uqtACevXxVVdlUNChJiUuPO5w
31iG66iwLWt8N0i8il0ShRvbmNEMFA26Tdvceu6XJW92p3W7XPkMF1Hh4iWtpM4mA3NsTZxOp+do
2niAj91KllySipUEVJnRACvSEs/z8zp8ZzPHDRjMry6NDxd3lz8BgLor/niTKDTEUskUxvtPxqer
O2kjg0SPMwRQ7MRlRZPrrSN2d3MM+evFtVxsY7yTLDLAh9JQG6gpdGj+6er2Yn5xdzOHxmNDZDEY
D9vaGFrpiiJqLR4Zi+1mszGciDEpf1uBsdisMGgtYnwdBaBiab5RNnlberTTX6FrQ7lXxM8VcMzn
iA6+lvpytyupB1DZbQJfshh3ML01i8E3t7Bg4DEsuevjbfC2gN7AQ1BBcjusA798fH8P4ioAwSpD
s1tXahw7lidSKAM+zXTYtFjAvbpSp5d0rUrMrDi5Jo9Y1RTxCg0keyH8f8KBc6wD7noHdFhDFiFG
+85Yilk5SSliRVOKVzojVR4OVp5tVlG5xOKdQ5PJovGG+oTuT7IUbbcyo5gZKY55cxg8WuEkW5o4
KjAmdaDrgHSW/s7akXKEGhkXpZiBa2VRsUEIDQnd2vBSSgx5LMsEWxpDqrjvXEd4S0PnFfV967BD
fi9L5glYlSSxJQ6nKEAajQVIYnkK0lH/UyCinjwE0EuD+3uBEU6nw5KanwpKehuewWWsC1zE8gSX
Z0OcAE1xKQt0srjktweY3OVkbEoeJ8Ijrugz2EyGAhuxVLF5xv95YJIZcfhS7W+a1OtUYHLzE1Gh
kXYclJE+amt9/FwnK4qSg+zJI/dsE2laW7gPshfKC0EoJDz5ImV4weMjmXssaY+zmTaUEYY8sZ0p
nEnL8x4pDnGbLcPdZ/QdxzC/q5/lJpISYkapREskVJ1aLuwJYWHU/yw4E52tCPoatMkIKRh9ZGvX
H4yr+Vx64QY41XFkCwvM4h2ioNzhWV9B4/u48Ufwog1Zu5+cS+1bJR8FCeREVadI12DFIvxsB2AJ
sPvieywqNNYEnxLLf66P2NNHTlhgqFfAGVXA8ZMLiO2FJSz5z6AKFlryJYR4fvGDrCwzCEJOHOSB
ReH/gqAgyHk2k0o2mkrpEIBjVSUAk2UfgNn42r3veI8Mx8uaeV8vv+79207GOBSfMGzcNigcvc5m
jRUx0trXA0R9WSbZy4yoa3rf6vVtvdMxneFQ14Y98wBRX1aJ+lIA0x/QP6j+IIVlu2JIiKtMZVrw
5NuffzQurq9vLokK9Xb48w+/3d/fC/I0oP9DJY3IHol3ibojEiI+0tdkNFEZ0n9pSQ14lA4q0/NC
y6CjCaJLp9aQkOOxtewTl5077XmjODpOxDLlSgZTTqvFGrC/Rdkx+hECPn0yWPlDpk4O2iSbY0yt
MPsHEmz214YPAAA=
EOF
)
```

**Comment by liviuchircu** [comment]
Mananged to repro easily, and the patch fixes the crash.  Backported to `3.2` as well.  Cheers!

**Comment by AdrianFretwell** [comment]
Liviu, I have been trying to test this all morning.  The patch fixes the SIGFPE, so we no longer get the restart.  But we do not detect any fraud.

My call to check_fraud, now returns ($rc) 1 (matching rule found) even if there is no matching rule.

Also if I do set a matching rule with a very low sequential or total call count, it never detects the calls going over the threshold.

I hope that makes sense, I will test further, but I'm sure something is wrong.

Adrian.

**Comment by liviuchircu** [comment]
Hi, thanks for the update!  Let me do some more elaborate tests then, I'm sure we'll get to the bottom of it :)

**Comment by AdrianFretwell** [comment]
Liviu,  I have just deleted all the rules from the fraud_detection table and restarted OpenSIPS.  Check_fraud() still sets the value of $rc to 1 event with no records in the table.

**Comment by liviuchircu** [comment]
@AdrianFretwell I did the following:

* used the exact rules from the [fraud_detection tutorial](https://www.opensips.org/Documentation/Tutorials-FraudDetection-3-1#toc4).  Since the current time matches `Mon-Fri` / 09:00 - 17:00`, the `check_fraud()` returns a `1` value, using profileid `1`.
* deleted all rules, reloaded with `opensips-cli -x mi fraud_reload`.  The `check_fraud()` immediately returns a `2` value, signifying that no rules were found, but there is no fraud trigger.

I'm not sure why you're not getting a `2` return code with an empty table.  Make sure you haven't cleared the wrong table or restarted the wrong OpenSIPS instance, etc. (that sort of problem)

Please open up a ticket with a minimal `opensips.cfg` + DB data (if possible!) that exhibits the fraud detection bug you're running into, so I know how to reproduce.

**Comment by AdrianFretwell** [comment]
Liviu, please accept my apologies. the $rc = 1 was being set by a call to xlog that I had added after the call to check_fraud to detect the SIGFPE.  So my fraud detection was testing the return from xlog not check_fraud!

It is very easy to confuse yourself with testing...

Many thanks for all your help.
Adrian.
